src/magic_folder/magicfolderdb.py

```diff
@@ -202,3 +214,47 @@ def did_upload_version(self, relpath_u, version, last_uploaded_uri, last_downloa
                                      relpath_u))
                 action.add_success_fields(insert_or_update=u"update")
             self.connection.commit()
+
+    def store_local_snapshot(self, serialized_snapshot, name):
+        """
+        Store or update the given serialized form of Local Snapshot for the
+        given the magicpath of the file (mangled file path).
+
+        :param str serialized_snashot: A JSON representation of a LocalSnapshot
+
+        :param str name: magicpath of the filepath whose snapshot is being stored
+        """
+        action = STORE_OR_UPDATE_SNAPSHOTS(
+            relpath=name,
+        )
+        with action:
+            try:
+                self.cursor.execute("INSERT INTO local_snapshots VALUES (?,?)",
+                                    (name, serialized_snapshot))
+                action.add_success_fields(insert_or_update=u"insert")
+            except (self.sqlite_module.IntegrityError, self.sqlite_module.OperationalError):
+                self.cursor.execute("UPDATE local_snapshots"
+                                    " SET snapshot_blob=?"
+                                    " WHERE path=?",
+                                    (serialized_snapshot, name))
+                action.add_success_fields(insert_or_update=u"update")
+            self.connection.commit()
```

Comment:
I'm afraid that an exception somewhere above will leave an open transaction on this connection.

These two links demonstrate what I think might be a safer pattern:

https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/ace36b198920dbeb745c7c82a9e8b4764d7af91a/src/_zkapauthorizer/model.py#L188-L201
https://github.com/PrivateStorageio/ZKAPAuthorizer/blob/ace36b198920dbeb745c7c82a9e8b4764d7af91a/src/_zkapauthorizer/model.py#L262-L283

Reply:
Thanks @exarkun. I will make another PR to fix the problem.